File: summary/go/12_dijkstra.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

type Graph map[string]map[string]float64

func dijkstra(graph Graph, source string) map[string]float64 {
	distances := make(map[string]float64)
	for vertex := range graph {
		distances[vertex] = math.Inf(1)
	}
	distances[source] = 0

	priorityQueue := make(PriorityQueue, 0)
	priorityQueue = append(priorityQueue, &Item{value: source, priority: 0})

	for len(priorityQueue) > 0 {
		currentVertex := heap.Pop(&priorityQueue).(*Item).value

		for neighbor, weight := range graph[currentVertex] {
			newDistance := distances[currentVertex] + weight
			if newDistance < distances[neighbor] {
				distances[neighbor] = newDistance
				heap.Push(&priorityQueue, &Item{value: neighbor, priority: newDistance})
			}
		}
	}

	return distances
}

type Item struct {
	value    string
	priority float64
	index    int
}

type PriorityQueue []*Item

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].priority < pq[j].priority
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*Item)
	item.index = n
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

func main() {
	graph := Graph{
		"A": {"B": 1, "C": 4},
		"B": {"A": 1, "C": 2, "D": 5},
		"C": {"A": 4, "B": 2, "D": 1},
		"D": {"B": 5, "C": 1},
	}
	distances := dijkstra(graph, "A")
	fmt.Println("Distances from A:", distances)
}

File: summary/c++/12_dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <limits>

using namespace std;

const int INF = numeric_limits<int>::max();

void dijkstra(const vector<vector<pair<int, int>>>& graph, int source, vector<int>& distances) {
    int n = graph.size();
    distances.assign(n, INF);
    distances[source] = 0;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, source});

    while (!pq.empty()) {
        int current_distance = pq.top().first;
        int current_vertex = pq.top().second;
        pq.pop();

        if (current_distance > distances[current_vertex]) {
            continue;
        }

        for (auto& edge : graph[current_vertex]) {
            int neighbor = edge.first;
            int weight = edge.second;
            int new_distance = current_distance + weight;

            if (new_distance < distances[neighbor]) {
                distances[neighbor] = new_distance;
                pq.push({new_distance, neighbor});
            }
        }
    }
}

int main() {
    int n = 4; // Number of vertices
    vector<vector<pair<int, int>>> graph(n);
    // Creating the graph:
    // A -> B (1), A -> C (4)
    graph[0].push_back({1, 1});
    graph[0].push_back({2, 4});
    // B -> A (1), B -> C (2), B -> D (5)
    graph[1].push_back({0, 1});
    graph[1].push_back({2, 2});
    graph[1].push_back({3, 5});
    // C -> A (4), C -> B (2), C -> D (1)
    graph[2].push_back({0, 4});
    graph[2].push_back({1, 2});
    graph[2].push_back({3, 1});
    // D -> B (5), D -> C (1)
    graph[3].push_back({1, 5});
    graph[3].push_back({2, 1});

    vector<int> distances;
    dijkstra(graph, 0, distances);

    for (int i = 0; i < n; i++) {
        cout << "Shortest distance from A to " << char('A' + i) << ": " << distances[i] << endl;
    }

    return 0;
}
